d2/src/main/java/com/linkedin/d2/balancer/subsetting/DeterministicSubsettingStrategy.java

```diff
@@ -0,0 +1,315 @@
+
+  /**
+   * Builds deterministic subsetting strategy
+   *
+   * @param metadataProvider  Provides peer cluster metadata information
+   * @param clusterName The name of the peer cluster
+   * @param minSubsetSize The minimum subset size to satisfy
+   */
+  public DeterministicSubsettingStrategy(DeterministicSubsettingMetadataProvider metadataProvider,
+                                         String clusterName,
+                                         int minSubsetSize)
+  {
+    _metadataProvider = metadataProvider;
+    MD5Hash hashFunction = new MD5Hash();
+    String[] keyTokens = {clusterName};
+    _randomSeed = new Random(hashFunction.hashLong(keyTokens));
+    _minSubsetSize = minSubsetSize;
+  }
+
+  @Override
+  public Map<T, Double> getWeightedSubset(Map<T, Double> pointsMap)
+  {
+    DeterministicSubsettingMetadata metadata = _metadataProvider.getSubsettingMetadata();
+    if (metadata != null)
+    {
+      if (metadata.equals(_currentMetadata))
+      {
+        if (_currentPointsMap != null && _currentWeightedSubset != null && !isPointsMapChanged(pointsMap))
+        {
+          return _currentWeightedSubset;
+        }
+        else
+        {
+          _currentPointsMap = pointsMap;
+        }
+      }
+      else
+      {
+        _currentMetadata = metadata;
+        _currentPointsMap = pointsMap;
+      }
+
+      List<T> points = new ArrayList<>(pointsMap.keySet());
+      Collections.shuffle(points, _randomSeed);
+      List<Double> weights = points.stream().map(pointsMap::get).collect(Collectors.toList());
+      Ring ring = new Ring(weights);
+
+      double offset = metadata.getInstanceId() / (double) metadata.getTotalInstanceCount();
+      double subsetSliceWidth = getSubsetSliceWidth(metadata.getTotalInstanceCount(), points.size());
+      List<Integer> indices = ring.getIndices(offset, subsetSliceWidth);
+
+      _currentWeightedSubset = indices.stream().collect(
+          Collectors.toMap(points::get, i -> round(ring.getWeight(i, offset, subsetSliceWidth), WEIGHT_DECIMAL_PLACE)));
+      return _currentWeightedSubset;
+    }
+    else
+    {
+      _log.warn("Cannot retrieve metadata required for D2 subsetting. Revert to use all available hosts.");
+      return null;
+    }
+  }
+
+  private static double round(double value, int places)
+  {
+    BigDecimal bd = new BigDecimal(Double.toString(value));
+    bd = bd.setScale(places, RoundingMode.HALF_UP);
+    return bd.doubleValue();
+  }
+
+  private static boolean isEqual(double a, double b, double delta)
+  {
+    return Math.abs(a - b) <= delta;
+  }
+
+  private boolean isPointsMapChanged(Map<T, Double> pointsMap)
+  {
+    if (pointsMap.size() != _currentPointsMap.size())
+    {
+      return true;
+    }
+
+    for (Map.Entry<T, Double> entry: pointsMap.entrySet())
+    {
+      if (!_currentPointsMap.containsKey(entry.getKey()) ||
+          !isEqual(_currentPointsMap.get(entry.getKey()), entry.getValue(), POINTS_MAP_CHANGED_DELTA))
```

Comment:
Agreed. I now uses the clusterGenerationId to check for staleness. It should cover the 2 cases you listed above. The clusterGenerationId is updated when URI/Cluster/Service/PartitionProperties are changed. And since cluster/service/partition properties are changed rarely, most of the times, we will regenerate subset map when URIs are updated.